File: LEETCODE/MEDIUM/DAY 25/JAVA SOLUTION/Top-Down Recursion with Bounds (O(n)).java
public boolean isValidBST(TreeNode root) {
    return valid(root, null, null);
}

private boolean valid(TreeNode node, Integer low, Integer high) {
    if (node == null) return true;
    if ((low != null && node.val <= low) || (high != null && node.val >= high))
        return false;
    return valid(node.left, low, node.val) && valid(node.right, node.val, high);
}

File: LEETCODE/MEDIUM/DAY 25/PYTHON SOLUTION/Top-Down Recursion with Bounds (O(n)).py
class Solution:
    def isValidBST(self, root):
        return self.valid(root, None, None)

    def valid(self, node, low, high):
        if not node:
            return True
        if (low is not None and node.val <= low) or (high is not None and node.val >= high):
            return False
        return self.valid(node.left, low, node.val) and self.valid(node.right, node.val, high)
